docs/rfc/rfc-015-abci++-tx-mutation.md

```diff
@@ -0,0 +1,261 @@
+
+- 23-Feb-2022: Initial draft (@williambanfield).
+- 28-Feb-2022: Revised draft (@williambanfield).
+
+## Abstract
+
+The specification of ABCI++ details a mechanism for proposers to replace transactions
+in the proposed block. This scheme requires the proposer to construct new transactions
+and mark these new transactions as replacing other removed transactions. The specification
+is somewhat ambiguous as to how the replacement may be communicated to peer nodes.
+This RFC discusses issues with this mechanism and possible solutions.
+
+## Background
+
+### What is the proposed change?
+
+The ABCI++ specification proposes mechanisms for adding, removing, and replacing
+transactions in a proposed block. To replace a transaction, the application running
+`ProcessProposal` can mark a transaction as replaced by other application-supplied
+transactions by returning a new transaction marked with the `ADDED` flag setting 
+the `new_hashes` field of the removed transaction to contain the list of transaction hashes
+that replace it. In the current specification for ABCI++, the full use of the
+`new_hashes` field is left somewhat ambiguous. At present, these hashes are not
+gossiped and are not eventually included in the block to signal replacement to
+other nodes. The specification does indicate that the transactions specified in
+the `new_hashes` field will be removed from the mempool but it's not clear how
+peer nodes will learn about them.
+
+### What systems would be affected by adding transaction replacement?
+
+The 'transaction' is a central building block of a Tendermint blockchain, so adding
+a mechanism for transaction replacement would require changes to many aspects of Tendermint.
+
+The following is a rough list of the functionality that this mechanism would affect:
+
+#### Transaction indexing
+
+Tendermint's indexer stores transactions and transaction results using the hash of the executed
+transaction [as the key][tx-result-index] and the ABCI results and transaction bytes as the value.
+
+To allow transaction replacement, the replaced transactions would need to stored as well in the
+indexer, likely as a mapping of original transaction to list of transaction hashes that replaced
+the original transaction.
+
+#### Transaction inclusion proofs
+
+The result of a transaction query includes a Merkle proof of the existence of the
+transaction in the block chain. This [proof is built][inclusion-proof] as a merkle tree
+of the hashes of all of the transactions in the block where the queried transaction was executed.
+
+To allow transaction replacement, these proofs would need to be updated to prove
+that a replaced transaction was included by replacement in the block.
+
+#### RPC-based transaction query parameters and results
+
+Tendermint's RPC allows clients to retrieve information about transactions via the
+`/tx_search` and `/tx` RPC endpoints.
+
+RPC query results containing replaced transactions would need to be updated to include
```

Comment:
I'm putting this comment here, but it applies in several places. 🙂 

I'm concerned that allowing the application to "replace" submitted transactions as part of the proposal process introduces a communication dependency between the submitter (e.g., a client on behalf of some user) and the application, that Tendermint cannot correctly mediate without knowing something about the semantics of the transaction for the application.

For example, suppose S submits a transaction that means "transfer X tokens from S to M". Even if we assume the application issues some kind of "identifier" for the transaction that is durable across changes, S has now a concern:

1. A Byzantine application instance could replace the transaction with "transfer X+5 tokens from S to M and transfer 200 tokens to StealyMcBurglar". As long as this is a structurally valid transaction, the rest of the network will vote for it and commit it to a block. By the time S discovers their identifier corresponds to a corruption of their original intent, it's too late.

2. On the other hand, some modifications are presumably fine: "transfer X tokens from S to T and also grant S a gold star" is different from what S proposed, but at least notionally unobjectionable.

To avoid the bad case (1), the submitter needs some way to check that the ID they got is a meaningful "certificate" of their intent. That's fundamentally a semantic property, and would vary by application. For some applications, maybe certain parts of the original transaction are "signed", and the ID proves to the submitter that those signed parts are unmodified. 

The point is: Tendermint can't really mediate this communication on behalf of the submitter. Whether we use hashes or application-assigned IDs, we basically have to trust that the relationship between the original transaction T and the modified transaction T' is "correct" in whatever way makes sense. Trying to keep track of that and prove it's valid isn't just a lot of work—it's unsound unless consensus also knows how to validate the semantic properties.

I don't know what a good solution here is, but my intuition is that Tendermint should avoid being in the business of tracking the association at all. As far as replicating the bits, it doesn't matter to Tendermint whether T or T' is adopted into the block. They are different bits, they might as well be different transactions. The fact that T' actually got injected by the application doesn't matter.

If we adopted the view that T and T' are different transactions, it's entirely up to the application to judge the contents of the pending transaction pool. If T' was the result of Byzantine application behaviour, the rest of the network won't vote for it, and it'll be purged in the next height. Meanwhile, the original submitted T remains for other nodes to consider (or drop) when it's their turn to propose.

How will S actually verify that their transaction gets committed? I think that question remains the same regardless how IDs are assigned: If an ID can correspond to different bits, there's a trust problem that can (I think) only be resolved by shared knowledge between the submitter and the application.

Reply:
If we want Tendermint to mediate the tx hash return, I think we'd have to make something seg-wit like. Where the application returns the bytes to Tendermint that are guaranteed to not change over prepare-proposal / any proving scheme. (A fundamental minimum to communicate the state change)

A first pass at this would imo just be excluding the signature from the tx hash. (And requiring via ProcessProposal that all txs that get on chain at minimum have a correct signature for the sequence number, and no sequence number replays) 

This gets you a large % of the desired goal for the near future. Seg-wit is fundamentally extending doing this.

Reply:
I would like for the answer here to be that Tendermint just copies the bits around, and isn't responsible for tracking the changing fortunes of a submitter's request.

However, we've set up the expectation that a submitter can track their bits through the system by their content address, and it seems like a lot of tools depend on that. Allowing PrepareProposal to change those bits in ways other than (perhaps) packaging them into groups not only breaks that expectation, it also introduces a bookkeeping problem for consensus that seems may require some kind of recursive shared mapping that is propagated among all the nodes in the network. 

Application assigned IDs may help, but I think that's not really the hard part of the problem. Does each node need to keep and update a two-way mapping of hashes to IDs? How is that mapping constructed and maintained to ensure we preserved the original data in case voting fails and it becomes our turn to propose?

I don't think this problem is infeasible, but the semantics of updating transactions seems wildly under-specified, and to a degree that worries me. I feel like we need clear properties to design against.

Reply:
My current leaning is that there really isn't much value to this kind of change. We could implement a system where only certain subsets of the request bytes are allowed to change, and where the application signals to Tendermint mutable vs immmutable sections of the transaction, and Tendermint, say, hashes only the immutable section. 

This is all reasonably _doable_, I guess I just am not sure I see the value. It might be there, but I'm not sure I totally see the reasoning for it as of now.

Reply:
I do not think that's something Tendermint should do at all. But by allowing the application to rewrite transactions arbitrarily, we can no longer make a meaningful guarantee that a transaction submitted to the consensus node will be correlated to what lands in the block. 

In principle the application has to provide that correlation, but the current API can't really be satisfied. Even if the node keeps a mapping between "application assigned ID X and bit strings T, T', T'' etc." all of that is extra-consensual. The user has no reason to trust that their transaction actually made it into the block unless they go around and scrape up all the possible realizations of that ID across all the nodes that may have acted as proposers during that height. 

At that point, I'd argue that the idea of inspecting the mempool to find a transaction becomes essentially worthless. 

Maybe we're okay with that, but a large chunk of the ecosystem is built on the ability to poke fingers into the mempool and follow the results. I haven't yet seen (or thought of) a real solution to that. It's like we'd need a meta consensus just to track IDs.

Reply:
I'm more saying that we could add some additional structure to our transactions that would allow the application to alter them without fundamentally breaking the authenticity of the transaction. If we created a structure like so:

```proto
message Transaction {
bytes body = 1;
bytes metadata = 2;
}
```

Where `body` was the only piece of data that Tendermint would guarantee was authentic, we could allow rewriting of the metadata section and perhaps even require it to be signed by the node that makes the change so as to ensure some level of authenticity. Users would always be able to track the address of the `body` field and the application could mangle the metadata to their hearts content. Both would wind up on-chain and would be as authentic as the systems that produced them allow. This is possible I'm just not sure it's all that useful.